cmd/private-prow-configs-mirror/main.go

```diff
@@ -0,0 +1,460 @@
+	}
+	return agent.Config(), nil
+}
+
+func updateProwConfig(configFile string, config prowconfig.ProwConfig) error {
+	data, err := yaml.Marshal(config)
+	if err != nil {
+		return fmt.Errorf("could not marshal Prow configuration: %v", err)
+	}
+
+	return ioutil.WriteFile(configFile, data, 0644)
+}
+
+func updateProwPlugins(pluginsFile string, config *plugins.Configuration) error {
+	data, err := yaml.Marshal(config)
+	if err != nil {
+		return fmt.Errorf("could not marshal Prow configuration: %v", err)
+	}
+
+	return ioutil.WriteFile(pluginsFile, data, 0644)
+}
+
+func privateOrgRepo(repo string) string {
+	return fmt.Sprintf("%s/%s", openshiftPrivOrg, repo)
+}
+
+func asYAML(v interface{}) string {
+	b, err := yaml.Marshal(v)
+	if err == nil {
+		return string(b)
+	}
+	return ""
+}
+
+func getOrgsWithOfficialImages() (orgsWithOfficialImages, error) {
+	ret := make(orgsWithOfficialImages)
+
+	callback := func(c *api.ReleaseBuildConfiguration, i *config.Info) error {
+		if !promotion.BuildsOfficialImages(c) {
+			return nil
+		}
+
+		if org, ok := ret[i.Org]; ok {
+			if repository, ok := org.repos[i.Repo]; ok {
+				repository.branches.Insert(i.Branch)
+				ret[i.Org].repos[i.Repo] = repository
+			} else {
+				ret[i.Org].repos[i.Repo] = repo{branches: sets.NewString(i.Branch)}
+			}
+		} else {
+			ret[i.Org] = orgWithOfficialImages{
+				repos: map[string]repo{i.Repo: {branches: sets.NewString(i.Branch)}},
+			}
+		}
+		return nil
+	}
+
+	if err := config.OperateOnCIOperatorConfigDir(config.CiopConfigInRepoPath, callback); err != nil {
+		return ret, fmt.Errorf("error while operating in ci-operator configuration files: %v", err)
+	}
+
+	return ret, nil
+}
+
+func injectPrivateBranchProtection(branchProtection prowconfig.BranchProtection, orgs orgsWithOfficialImages) {
+	logger := logrus.WithField("component", "config.branch-protection")
+	privateOrg := prowconfig.Org{
+		Repos: make(map[string]prowconfig.Repo),
+	}
+
+	logger.Info("Processing...")
+	for orgName, orgValue := range branchProtection.Orgs {
+		if _, ok := orgs[orgName]; ok {
+			for repoName, repoValue := range orgValue.Repos {
+				if _, ok := orgs[orgName].repos[repoName]; ok {
+					logger.WithField("repo", repoName).Info("Found")
+					privateOrg.Repos[repoName] = repoValue
+				}
+			}
+		}
+	}
+
+	if len(privateOrg.Repos) > 0 {
+		logger.WithField("value", asYAML(privateOrg)).Info("Generating openshift-priv org.")
+		branchProtection.Orgs[openshiftPrivOrg] = privateOrg
+	}
+}
+
+func injectPrivateTideOrgContextPolicy(contextOptions prowconfig.TideContextPolicyOptions, orgs orgsWithOfficialImages) {
+	logger := logrus.WithField("component", "config.context_options")
+	logger.Info("Processing...")
+	privateOrgRepos := make(map[string]prowconfig.TideRepoContextPolicy)
+	for orgName, orgValue := range contextOptions.Orgs {
+		if _, ok := orgs[orgName]; ok {
+			for repoName, repoValue := range orgValue.Repos {
+				if _, ok := orgs[orgName].repos[repoName]; ok {
```

Comment:
Better yet, write your types to describe what they are going to do. My earlier suggestions were based on your names, which made it sound like you needed to know the list of images that were promoted. As of yet I'm seeing this access pattern commonly, so you want a type that makes simple and easy to answer the question "is this org/repo(branch) promoting and do I therefore need to care about it?"

Reply:
So, for instance, if you had:

```go
type officialBranchesByOrgRepo map[string]map[string]sets.String
```

You could have 

```go
func (o *officialBranchesByOrgRepo) isOfficialRepo(org, repo string) bool {
    if repos, ok := o[org]; ok {
        _, isOfficial := o[org]
        return isOfficial
    }
    return false
}

func (o *officialBranchesByOrgRepo) isOfficialBranch(org, repo, branch string) bool {
    if ok := o.isOfficialRepo(org,repo); ok  {
        _, isOfficial := o[org][repo][branch]
        return isOfficial
    }
    return false
}
```